File: even-hub-display/src/models/RideData.ts
/**
 * Lifecycle of a ride, so the display knows when to stop showing a card.
 *
 * Without this the last ride stayed on screen forever — a HUD confidently
 * reporting "1 MIN · ARRIVING" long after the trip ended.
 */
export type RideStatus =
  | 'enroute'    // driver on the way, ETA meaningful
  | 'arriving'   // pulling up now
  | 'arrived'    // waiting at the pickup point
  | 'completed'  // trip finished
  | 'cancelled'  // ride cancelled

/** Statuses after which the card should be cleared rather than kept alive. */
export const TERMINAL_RIDE_STATUSES: readonly RideStatus[] = [
  'arrived',
  'completed',
  'cancelled'
]

export function isTerminalStatus(status?: RideStatus): boolean {
  return status !== undefined && TERMINAL_RIDE_STATUSES.includes(status)
}

/**
 * RideData model matching the iOS companion app + advanced features
 */
export interface RideData {
  /** Defaults to 'enroute' when a producer omits it. */
  status?: RideStatus
  driverName: string
  driverRating: number
  vehicleMake: string
  vehicleModel: string
  vehicleColor: string
  licensePlate: string
  etaMinutes: number
  driverLat?: number
  driverLng?: number
  requesterLat?: number
  requesterLng?: number
  timestamp: string
  // Advanced features
  driverPhone?: string
  driverAvatar?: string
  vehiclePhoto?: string
  driverFullName?: string
  totalRides?: number
  acceptanceRate?: number
  cancellationRate?: number
}

/**
 * Chat message interface for driver communication
 */
export interface ChatMessage {
  id: string
  sender: 'driver' | 'user'
  message: string
  timestamp: Date
  read: boolean
}

/**
 * API request/response types
 */
export interface RideUpdateRequest extends RideData {
  // Same structure as RideData
}

export interface ApiResponse<T> {
  status: string
  data?: T
  error?: string
  message?: string
}

export interface HealthResponse {
  status: string
  uptime?: number
  timestamp?: string
}

File: ios-companion/RideData.swift
import Foundation

/// Data model for a ride notification parsed from Uber
struct RideData: Codable {
    let driverName: String
    let driverRating: Double
    let vehicleMake: String
    let vehicleModel: String
    let vehicleColor: String
    let licensePlate: String
    let etaMinutes: Int
    let driverLat: Double?
    let driverLng: Double?
    let requesterLat: Double?
    let requesterLng: Double?
    let timestamp: Date

    enum CodingKeys: String, CodingKey {
        case driverName
        case driverRating
        case vehicleMake
        case vehicleModel
        case vehicleColor
        case licensePlate
        case etaMinutes
        case driverLat
        case driverLng
        case requesterLat
        case requesterLng
        case timestamp
    }

    /// Initialize with default values for optional fields
    init(
        driverName: String,
        driverRating: Double = 0.0,
        vehicleMake: String = "",
        vehicleModel: String = "",
        vehicleColor: String = "",
        licensePlate: String = "",
        etaMinutes: Int = 0,
        driverLat: Double? = nil,
        driverLng: Double? = nil,
        requesterLat: Double? = nil,
        requesterLng: Double? = nil,
        timestamp: Date = Date()
    ) {
        self.driverName = driverName
        self.driverRating = driverRating
        self.vehicleMake = vehicleMake
        self.vehicleModel = vehicleModel
        self.vehicleColor = vehicleColor
        self.licensePlate = licensePlate
        self.etaMinutes = etaMinutes
        self.driverLat = driverLat
        self.driverLng = driverLng
        self.requesterLat = requesterLat
        self.requesterLng = requesterLng
        self.timestamp = timestamp
    }
}
